src/main/java/org/elasticsearch/gateway/GatewayMetaState.java

```diff
@@ -225,9 +218,116 @@ private void pre20Upgrade() throws Exception {
                 }
             }
         }
-        if (hasCustomPre20HashFunction|| pre20UseType != null) {
+        if (hasCustomPre20HashFunction || pre20UseType != null) {
             logger.warn("Settings [{}] and [{}] are deprecated. Index settings from your old indices have been updated to record the fact that they "
                     + "used some custom routing logic, you can now remove these settings from your `elasticsearch.yml` file", DEPRECATED_SETTING_ROUTING_HASH_FUNCTION, DEPRECATED_SETTING_ROUTING_USE_TYPE);
         }
     }
+
+    /**
+     * Loads the current meta state for each index in the new cluster state and checks if it has to be persisted.
+     * Each index state that should be written to disk will be returned. This is only run for data only nodes.
+     * It will return only the states for indices that actually have a shard allocated on the current node.
+     *
+     * @param event           the cluster state event from which we figure out what is new in each index and should potentially be written
+     * @param currentMetaData the current index state in memory.
+     * @return iterable over all indices states that should be written to disk
+     */
+    public static Iterable<GatewayMetaState.IndexMetaWriteInfo> filterStateOnDataNode(ClusterChangedEvent event, MetaData currentMetaData, MetaStateService metaStateService, ESLogger logger) {
+        Map<String, GatewayMetaState.IndexMetaWriteInfo> indicesToWrite = new HashMap<>();
+        RoutingNode thisNode = event.state().getRoutingNodes().node(event.state().nodes().localNodeId());
+        if (thisNode == null) {
+            // this needs some other handling
+            return indicesToWrite.values();
+        }
+        // iterate over all shards allocated on this node in the new cluster state but only write if ...
+        for (MutableShardRouting shardRouting : thisNode) {
+            IndexMetaData indexMetaData = event.state().metaData().index(shardRouting.index());
+            IndexMetaData currentIndexMetaData = maybeLoadIndexState(currentMetaData, indexMetaData, metaStateService, logger);
+            String writeReason = null;
+            // ... state persistence was disabled or index was newly created
+            if (currentIndexMetaData == null) {
+                writeReason = "freshly created";
+                // ... new shard is allocated on node (we could optimize here and make sure only written once and not for each shard per index -> do later)
```

Comment:
see below. check now if shard was allocated on node in previous cluster state and if not do not write.